docs/cookbook/named_entity_recognition.md

```diff
@@ -0,0 +1,269 @@
+
+    Now, analyze the following text:
+
+    Text: {text}
+    """
+)
+def improved_ner(text: str): ...
+
+complex_text = """
+Apple Inc., the tech giant founded by Steve Jobs and Steve Wozniak, recently announced a partnership with OpenAI, the artificial intelligence research laboratory consisting of the for-profit corporation OpenAI LP and its parent company, the non-profit OpenAI Inc. This collaboration aims to enhance Siri, Apple's virtual assistant, which competes with Amazon's Alexa and Google Assistant, a product of Alphabet Inc.'s Google division. The joint project will be led by Apple's AI chief John Giannandrea, a former Google executive, and will take place at Apple Park, the company's headquarters in Cupertino, California.
+"""
+print("\nImproved NER Results:")
+improved_result = improved_ner(complex_text)
+
+def print_nested_entities(entities, level=0):
+    for entity in entities:
+        indent = "  " * level
+        entity_info = f"Entity: {entity.entity}, Label: {entity.label}, Parent: {entity.parent}"
+        print(textwrap.indent(entity_info, indent))
+        if entity.children:
+            print_nested_entities(entity.children, level + 1)
+
+print_nested_entities(improved_result.entities)
+
+# Output
+"""
+Improved NER Results:
+Entity: Apple Inc., Label: ORGANIZATION, Parent: None
+  Entity: Steve Jobs, Label: PERSON, Parent: Apple Inc.
+  Entity: Steve Wozniak, Label: PERSON, Parent: Apple Inc.
+  Entity: OpenAI, Label: ORGANIZATION, Parent: Apple Inc.
+    Entity: OpenAI LP, Label: ORGANIZATION, Parent: OpenAI
+      Entity: OpenAI Inc., Label: ORGANIZATION, Parent: OpenAI LP
+  Entity: John Giannandrea, Label: PERSON, Parent: Apple Inc.
+  Entity: Apple Park, Label: LOCATION, Parent: Apple Inc.
+  Entity: Cupertino, Label: LOCATION, Parent: Apple Park
+  Entity: California, Label: LOCATION, Parent: Cupertino
+Entity: OpenAI LP, Label: ORGANIZATION, Parent: OpenAI
+Entity: OpenAI Inc., Label: ORGANIZATION, Parent: OpenAI LP
+Entity: Alphabet Inc., Label: ORGANIZATION, Parent: None
+  Entity: Google, Label: ORGANIZATION, Parent: Alphabet Inc.
+    Entity: Google Assistant, Label: ORGANIZATION, Parent: Google
+Entity: Amazon, Label: ORGANIZATION, Parent: None
+  Entity: Alexa, Label: ORGANIZATION, Parent: Amazon
+"""
+```
+
+## Utility Functions
+
+To help visualize the hierarchical structure of entities, we can use a recursive printing function:
+
+```python
+import textwrap
+
+def print_entity_tree(entity, level=0):
+    indent = "  " * level
+    entity_info = f"Entity: {entity.entity}, Label: {entity.label}, Parent: {entity.parent}"
+    print(textwrap.indent(entity_info, indent))
+    for child in entity.children:
+        print_entity_tree(child, level + 1)
+```
+
+## Testing
+
+To ensure robustness, it's crucial to test the NER system with diverse scenarios. Here's a function to run multiple test cases:
+
+```python
+def test_ner_system(recognize_entities_func):
```

Comment:
this test should actually test the input with an expected output rather than just printing things out

Reply:
I didnt understand how to validate it. Any particular way to validate them ?

Reply:
the goal would be to have input text with a known output NER structure, run the NER against the input and then confirm the output matches the expected output.